### PR TITLE
Improve decimal parsing performance

### PR DESCRIPTION
**Pull request by spebern**
# Which issue does this PR close?

# Rationale for this change
 
Improve decimal parsing performance (~50%)

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by tustvold** [comment]
Thank you for this, will review fully tomorrow. Just some ideas that might make this even faster

* Handle the sign, if any, outside the loop
* Split the string on a decimal point, if any, so the two parts can be processed using the optimised lexical_core integer parsing routines

**Comment by spebern** [comment]
Thanks for the suggestions. I will have a look the next few days!

**Comment by spebern** [comment]
```
   Compiling arrow-cast v35.0.0 (/home/ben/workspace/arrow-rs/arrow-cast)
    Finished bench [optimized] target(s) in 12.64s
     Running benches/parse_decimal.rs (/home/ben/workspace/arrow-rs/target/release/deps/parse_decimal-ed9b69f81fe2bf18)
123.123                 time:   [104.52 ns 104.72 ns 104.95 ns]
                        change: [-52.561% -52.471% -52.383%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

123.1234                time:   [104.07 ns 104.20 ns 104.36 ns]
                        change: [-53.401% -53.291% -53.181%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  7 (7.00%) high mild
  3 (3.00%) high severe

123.1                   time:   [98.928 ns 99.199 ns 99.470 ns]
                        change: [-45.234% -45.059% -44.861%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe

123                     time:   [91.060 ns 91.155 ns 91.230 ns]
                        change: [-39.558% -39.512% -39.470%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  1 (1.00%) low severe
  3 (3.00%) low mild
  2 (2.00%) high mild

-123.123                time:   [105.71 ns 105.81 ns 105.91 ns]
                        change: [-52.052% -51.944% -51.796%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  3 (3.00%) high mild
  1 (1.00%) high severe

-123.1234               time:   [106.38 ns 106.46 ns 106.53 ns]
                        change: [-51.522% -51.477% -51.428%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe

-123.1                  time:   [101.36 ns 101.52 ns 101.70 ns]
                        change: [-44.430% -44.366% -44.301%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 17 outliers among 100 measurements (17.00%)
  2 (2.00%) low severe
  14 (14.00%) low mild
  1 (1.00%) high mild

-123                    time:   [93.470 ns 93.481 ns 93.493 ns]
                        change: [-38.446% -38.425% -38.404%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) high mild
  2 (2.00%) high severe

0.0000123               time:   [76.150 ns 76.335 ns 76.565 ns]
                        change: [-50.658% -50.451% -50.165%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  2 (2.00%) low mild
  1 (1.00%) high mild
  2 (2.00%) high severe

12.                     time:   [75.427 ns 75.511 ns 75.606 ns]
                        change: [-38.211% -38.128% -38.050%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 15 outliers among 100 measurements (15.00%)
  7 (7.00%) high mild
  8 (8.00%) high severe

-12.                    time:   [77.339 ns 77.507 ns 77.684 ns]
                        change: [-36.242% -36.182% -36.113%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  5 (5.00%) high mild
  6 (6.00%) high severe

00.1                    time:   [84.054 ns 84.180 ns 84.291 ns]
                        change: [-42.095% -42.014% -41.947%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 23 outliers among 100 measurements (23.00%)
  12 (12.00%) low severe
  4 (4.00%) low mild
  7 (7.00%) high mild

-00.1                   time:   [85.438 ns 85.652 ns 85.869 ns]
                        change: [-43.774% -43.677% -43.580%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 32 outliers among 100 measurements (32.00%)
  8 (8.00%) low severe
  5 (5.00%) low mild
  5 (5.00%) high mild
  14 (14.00%) high severe

12345678912345678.1234  time:   [320.02 ns 320.41 ns 320.75 ns]
                        change: [-52.900% -52.848% -52.797%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  4 (4.00%) low severe
  8 (8.00%) low mild
  1 (1.00%) high mild

-12345678912345678.1234 time:   [321.51 ns 321.92 ns 322.31 ns]
                        change: [-52.910% -52.852% -52.796%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) low mild
  1 (1.00%) high severe

99999999999999999.999   time:   [321.10 ns 321.52 ns 321.86 ns]
                        change: [-53.848% -53.781% -53.718%] (p = 0.00 < 0.05)
                        Performance has improved.

-99999999999999999.999  time:   [321.45 ns 321.78 ns 322.10 ns]
                        change: [-54.293% -54.246% -54.201%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  7 (7.00%) low mild
  1 (1.00%) high mild

.123                    time:   [59.960 ns 60.020 ns 60.086 ns]
                        change: [-50.726% -50.672% -50.624%] (p = 0.00 < 0.05)
                        Performance has improved.

-.123                   time:   [60.894 ns 60.979 ns 61.080 ns]
                        change: [-50.659% -50.555% -50.477%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 17 outliers among 100 measurements (17.00%)
  3 (3.00%) low severe
  2 (2.00%) low mild
  3 (3.00%) high mild
  9 (9.00%) high severe

123.                    time:   [90.291 ns 90.504 ns 90.719 ns]
                        change: [-41.874% -41.739% -41.605%] (p = 0.00 < 0.05)
                        Performance has improved.

-123.                   time:   [92.033 ns 92.224 ns 92.414 ns]
                        change: [-40.311% -40.201% -40.085%] (p = 0.00 < 0.05)
                        Performance has improved.

     Running benches/parse_timestamp.rs (/home/ben/workspace/arrow-rs/target/release/deps/parse_timestamp-4e4733f613306eeb)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

**Comment by spebern** [comment]
I had a look at what @tustvold suggested and in some cases we'd get another ~40% of performance. However, I only tested this for the 128 bit types, because `lexical` doesn't support 256 bit. I haven't fully grasped yet how to implement `lexical` support for the 256 bit types here yet, but basic things such as shift operations are missing, so it will be a little more work.

What do you think of replacing the usage of `BigInt` by https://docs.rs/ethnum/latest/ethnum/ ? `ethnum` is also used by arrow2 and has a more complete implementation when it comes to basic integer operations such as bit shifting.

**Comment by tustvold** [comment]
> basic integer operations such as bit shifting

These should be relatively straightforward to implement, I will see what I can do

**Comment by spebern** [comment]
Does it make sense to have a look at `ethnum` as a replacement? `arrow2`'s `i256` is a simple wrapper around it: https://github.com/jorgecarleitao/arrow2/blob/db87f71f3f673b240de80d5a7592d45c0beae805/src/types/native.rs#LL519-L519C35

Would probably save some work, since all the ops from the rust std lib seem to be implemented there.

I will have a look at `lexical` support this weekend, not quite sure for now how much is missing.

**Comment by tustvold** [comment]
> Does it make sense to have a look at ethnum as a replacement

I think if we can avoid a dependency, especially one that doesn't have a clear maintenance story (it is a personal crate), that is better. Let me see what it looks like, I think it should be straightforward.

**Comment by alamb** [comment]
> I think if we can avoid a dependency, especially one that doesn't have a clear maintenance story (it is a personal crate), that is better. Let me see what it looks like, I think it should be straightforward.

Related https://github.com/apache/arrow-rs/pull/3884

**Comment by alamb** [comment]
@spebern  would it be possible to update this PR to use the operations added in https://github.com/apache/arrow-rs/pull/3884 rather than a new crate?

**Review by tustvold — APPROVED** [review]
I think as is this PR represents a significant performance improvement, we can always further improve the performance in subsequent PRs. Thank you for this

File: arrow-cast/src/parse.rs [hunk 1]
```diff
@@ -613,100 +613,93 @@ pub fn parse_decimal<T: DecimalType>(
...
+                    fractionals += 1;
                 }
-                seen_sign = true;
+                digits += 1;
+                if digits > precision {
+                    return Err(ArrowError::ParseError(
+                        "parse decimal overflow".to_string(),
+                    ));
+                }
+                result = result.mul_checked(base)?;
```
Comment (tustvold) [inline arrow-cast/src/parse.rs:647]: It occurs to me that the precision check above should mean this can't overflow

**Pull request merged**